src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp

```diff
@@ -89,80 +89,84 @@ void CgroupV1MemoryController::set_subsystem_path(char *cgroup_path) {
 }
 
 jlong CgroupV1Subsystem::read_memory_limit_in_bytes() {
-  GET_CONTAINER_INFO(julong, _memory->controller(), "/memory.limit_in_bytes",
-                     "Memory Limit is: " JULONG_FORMAT, JULONG_FORMAT, memlimit);
-
-  if (memlimit >= os::Linux::physical_memory()) {
+  GET_CONTAINER_INFO(jlong, _memory->controller(), "/memory.limit_in_bytes",
+                     "Memory Limit is: " JLONG_FORMAT, JLONG_FORMAT, memlimit);
+  assert(memlimit > 0, "invariant");
```

Comment:
The thing with `GET_CONTAINER_INFO` is that it's a preprocessor macro which might trigger a short return. So this cannot happen (currently). The assert is a safeguard. I'm not sure we need this handled in product code. Thoughts?

Reply:
Surely it can happen because `subsystem_file_line_contents` may successfully parse `-1` into a `jlong`? Assuming that the file can contain a negative number. I don't think I understand why we have the safeguard there.

Reply:
The files won't contain negative numbers, AFAIK. Only some random large numbers or the actual limit. `-2` is being returned by the macro if the file doesn't exist or some other error occurs (basically this case). Since this repeatedly causes confusion for me when I revisit this code, I've added the assert for clarity. When we reach line 94, we can be sure that we get a positive value. If you want the assert removed, that's fine with me as well. To me, having the assert there is clearer, though.

Reply:
I think the `assert(memswlimit > 0, "invariant");` doesn't give a clear explanation why this is the case. And seeing multiple of this kind of assert in the same file is problematic.

Is it possible for `/memory.limit_in_bytes` to use a number that's larger than `max_jlong`? If so, `GET_CONTAINER_INFO(jlong, ... JLONG_FORMAT)` will fail and this function will return `OSCONTAINER_ERROR`. Previously, this function would return -1.

(the fact that `OSCONTAINER_ERROR` (-2) can be returned by this function is also problematic).

I actually think the old code is better. It can be further tightened by changing the blind (and potentially unsafe) cast

```
 return (jlong)hier_memswlimit;
```

to

```
 return as_positive_memory_limit(hier_memswlimit);
```

as_positive_memory_limit will assert that the value is not greater than `max_jlong`, which should be guaranteed by the above calculation. More explanations can be added as comments in `as_positive_memory_limit`.

Reply:
>-2 is being returned by the macro if the file doesn't exist or some other error occurs (basically this case).

Yes, but as you said it'll trigger a short return and never set `memlimit` if it fails, so we'll never see `-2` in `memlimit` regardless.

Now I understand your reasoning, and I get the idea. I do think that if you don't know the macro you'll be confused in one way with the assert, and if you do know the macro you'll be confused without the assert :-). The "real" answer is probably to fix these macros, I think we can do it better. Of course, future RFE.

You read this code more often than I do, I defer to you on keeping or not keeping the `assert`s.  Ping me when you've fixed the log issue please and I'll approve the PR.

Reply:
> I think the `assert(memswlimit > 0, "invariant");` doesn't give a clear explanation why this is the case. And seeing multiple of this kind of assert in the same file is problematic.

OK, I'll remove the asserts.

> Is it possible for `/memory.limit_in_bytes` to use a number that's larger than `max_jlong`?

Theoretically yes. This code really is accounting for `not-a-fixed-unlimited` value being specified for cgroups v1. So I argue we don't care. `max_jlong` is `9,223,372,036,854,775,807`. `max_julong` is `18,446,744,073,709,551,615`. Both are beyond *exabyte* values and we do compare against the actual physical memory of the system (which I'd argue won't be beyond exabytes any time soon?) anyway. We are dealing with a memory limit here, which are supposed to be *smaller* than the physical host values (common case).

> If so, `GET_CONTAINER_INFO(jlong, ... JLONG_FORMAT)` will fail and this function will return `OSCONTAINER_ERROR`. Previously, this function would return -1.
> 
> (the fact that `OSCONTAINER_ERROR` (-2) can be returned by this function is also problematic).

We don't care in this case, though. All negative values returned are being treated as "no limit". I.e. in both cases the container limit will be unlimited and there is no change from a users' perspective in that regard.

This is an attempt to fixing log traces, which the user actually *can* see and they don't really make sense, currently. For the theoretical case of the value exceeding `max_jlong` and then returning `OS_CONTAINER_ERROR` we are fine. The physical host memory value will be used instead. Even so, the trace would indicate `-2` in the logs indicating this. It's also not a question of correctness since we can assume that if the value is greater than `2^63-1` we have a unlimited config, unless there are really beyond exabytes memory containers out there.

> I actually think the old code is better. 

The old code wanted to deal with "recognizing unlimited container memory values". This is solved - since [JDK-8292083](https://bugs.openjdk.org/browse/JDK-8292083), by using the physical machine's memory as an upper bound of that value. There is really no point to run containers with a memory limit *larger* than the physical host. So I have to disagree. The old code of [JDK-8194232](https://bugs.openjdk.org/browse/JDK-8194232) was a kludge. The new code (with the physical's machine memory as an upper bound) is more sound.

Cases beyond `max_jlong` values are theoretical at best, and won't make a difference in this code. Once we have a large positive number, we know what to do.

Reply:
> I argue we don't care. max_jlong is 9,223,372,036,854,775,807

We do care if such a value can possibly be encountered. We don't want the VM to behave incorrectly. With the current implementation, it's hard to understand what happens when such a large value is encountered.

(In fact, the new behavior is wrong - the function will return -2 without considering `hier_memlimit`).

The problem is there's too much mental exercise when reading the code. There's casting, overflow/underflow, handling of special values of -1 and -2, and considerations for logging. Also, the same logic is duplicated in several places, so they can easily get out of sync in the future.

Is it possible to localize the logic to a utility function? 

Maybe something like this? All the casting and logging, etc, will be handled inside, in a single place.

```
// Returns -1 if the limit is not specified, or there was a failure when reading the limit, or if the specified
// limit is the same or larger than MIN2(physical_max, jlong_max).
// Otherwise returns a non-negative value that's smaller than MIN2(physical_max, jlong_max).
jlong limit = read_memory_limit("/memory.memsw.limit_in_bytes", physical_max);
```

`physical_max` will be a `julong`:

```
julong physical_max = os::Linux::physical_memory(); 
- or -
julong physical_max = os::Linux::host_swap() + os::Linux::physical_memory();
```

Reply:
Fair enough. I'll refactor the code into a separate function as you suggest. Thanks for your input!

Reply:
@iklam After some thought, moving to a utility would need to move it into a utility preprocessor `macro` since it uses `GET_CONTAINER_INFO` macros. Would you prefer that over the current version? Or is that not the way to go? Just checking..

Reply:
My suggestion is to first fix the `GET_CONTAINER_INFO` macros so that the logging works for julong types. That would address the main concern of this PR. How about this?

https://github.com/openjdk/jdk/compare/master...iklam:jdk:temp-improve-get-container-info-macros?expand=1

(the above changeset also simplifies `GET_CONTAINER_INFO_CPTR` so you don't need to specify the string length twice).

After that, if you want to consolidate the current logic to an utility function in a separate PR, that's fine too, but I think the current logic is correct (although a bit verbose).

Reply:
OK.